### PR TITLE
feat: Added endpoint to export all account balances

### PR DESCRIPTION
**Pull request by javiersuweijie**
#### Added new endpoint `POST /export/accounts` 

Exports a full list of accounts addresses, their vested and vesting luna. format of the export file is `account,vested,vesting` eg:
```
terra1qqqqqqqqqqqqqqqqqqqqqqqqqqqqqqqq486l9a,706511,1648524
terra1qqqz5f2l5epfft7l08wvp8qddhtapk437fy07t,931644922,2173312913
terra1qqqyhwh7q970w75ct3gf566en29yptyq6yhhgp,900000,296208964
terra1qqqyc5dkra4pnakgdx73zm57r0twekqtdg47wg,0,0
```

Exports only vesting and base accounts. Module accounts are excluded from the export. As of this PR:
```
total vesting + vested - module accounts = 674079642.260970
total vested (circulating supply) = 125850903.963203
```

### CODE REVIEW

**Comment by jeffwoooo** [comment]
@javiersuweijie How long does it take to iterate all accounts? Looks like we should call that for every block. (cc: @Vritra4)

**Comment by jeffwoooo** [comment]
I think if we can get accounts those amount changed in current block, we would make it faster.

@Vritra4 Do you have idea how we can get only changed accounts?

**Comment by etienne-napoleone** [comment]
Shouldn't we first consider if mantlemint should diverge from the LCD API?
I feel like this should be a different product (based on mantlemint)

**Comment by javiersuweijie** [comment]
> @javiersuweijie How long does it take to iterate all accounts? Looks like we should call that for every block. (cc: @Vritra4)

Currently it takes around 5s to iterate through all accounts. Calling for every block seems excessive since the circulating supply doesn't change much.  

> I think if we can get accounts those amount changed in current block, we would make it faster.

Couldn't really think of a way to do that right now since vesting accounts take in the block timestamp to returns how much is vesting. Full iteration is needed to get the total vesting amount.

**Comment by javiersuweijie** [comment]
> Shouldn't we first consider if mantlemint should or shouldn't diverge from the LCD API? I feel like this should be a different product/repo (based on mantlemint)

Agree that adding this functionality moves mantlemint away from being just a faster LCD. In the current state, it already has the `indexing` feature that is not part of the standard LCD interface. Since this feature is urgently needed by coinmarketcap and coingecko, what I'll do is to add a configuration to disable this feature by default so anyone who doesn't want it doesn't need to run it. 

@etienne-napoleone In parallel, let's start a conversation on setting a roadmap for mantlemint and defining what it should or should not achieve.

File: export/export.go [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	vestingtypes "github.com/cosmos/cosmos-sdk/x/auth/vesting/types"
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	"github.com/cosmos/cosmos-sdk/x/distribution/types"
+	distrotypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
+	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
+	terra "github.com/terra-money/core/v2/app"
+)
+
+var IsAccountExportRunning = false
```
Comment (hanjukim) [inline export/export.go:19]: Isn't this a local module variable? and I think `isAccountExportWorkerRunning` is more clearer

File: export/export.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+	communityPool, err := app.DistrKeeper.CommunityPool(sdktypes.WrapSDKContext(ctx), &distrotypes.QueryCommunityPoolRequest{})
+	if err != nil {
+		return sdktypes.Int{}, err
+	}
+	lunaCommunityPool := communityPool.Pool.AmountOf("uluna").TruncateInt()
+
+	return lunaTotalSupply.Sub(lunaCommunityPool).Sub(totalVesting), nil
+}
+
+func exportWorker(app *terra.TerraApp) {
```
Comment (hanjukim) [inline export/export.go:59]: `runAccountExportWorker` would be more clearer

File: rpc/register.go [hunk 3]
```diff
@@ -90,7 +99,7 @@ func StartRPC(
 	// caching middleware
 	apiSrv.Router.Use(func(next http.Handler) http.Handler {
 		return http.HandlerFunc(func(writer http.ResponseWriter, request *http.Request) {
-			if request.URL.Path == "/health" {
+			if request.URL.Path == "/health" || !export.ShouldCacheRoute(request.URL.Path) {
```
Comment (hanjukim) [inline rpc/register.go:102]: I think we need a better structure (scheme?) for handling custom endpoints

**Review by hanjukim — CHANGES_REQUESTED** [review]

File: export/client.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import (
+	"net/http"
+	"strings"
+
+	"github.com/gorilla/mux"
+	terra "github.com/terra-money/core/v2/app"
+)
+
+func ShouldCacheRoute(route string) bool {
```
Comment (hanjukim) [inline export/client.go:11]: Why did you name this function like this? I guess `ShouldHandleRoute` is more proper..?

**Review by hanjukim — APPROVED** [review]

File: export/export.go (hunk 1 above)
Comment (javiersuweijie) [reply export/export.go:19]: +1

File: export/export.go (hunk 2 above)
Comment (javiersuweijie) [reply export/export.go:59]: +1

File: export/client.go (hunk 4 above)
Comment (javiersuweijie) [reply export/client.go]: Currently all endpoints expect `/health` is cached and I wanted to exclude exports from the cache. But thinking through it, there is no reason not to cache `circulating_supply`. Let me remove this.

**Comment by etienne-napoleone** [comment]
> Agree that adding this functionality moves mantlemint away from being just a faster LCD. In the current state, it already has the `indexing` feature that is not part of the standard LCD interface.

Sorry what do you mean by the `indexing` feature? I might not know about those endpoints

> Since this feature is urgently needed by coinmarketcap and coingecko, what I'll do is to add a configuration to disable this feature by default so anyone who doesn't want it doesn't need to run it.

Separating those in another repo/fork doesn't make it shorter or longer to ship

> @etienne-napoleone In parallel, I think we can start a roadmap for mantlemint and defining what it should or should not achieve. Could add you into to the discussion if you want.

Would be interested thanks! With Setten we're probably going to be one of the largest mantlemint user in term of traffic served, so we'd be happy if we could have a voice

File: export/export.go [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		delegationRewards, err := app.DistrKeeper.DelegationTotalRewards(sdktypes.WrapSDKContext(ctx), &types.QueryDelegationTotalRewardsRequest{
+			DelegatorAddress: account.GetAddress().String(),
+		})
+		if err != nil {
+			panic(err)
+		} else {
+			balance = balance.Add(delegationRewards.Total.AmountOf("uluna").TruncateInt())
+		}
+		switch account.(type) {
+		case *vestingtypes.PeriodicVestingAccount:
```
Comment (jeffwoooo) [inline export/export.go:78]: Can we skip other Vesting Account Types?

**Comment by jeffwoooo** [comment]
> > Agree that adding this functionality moves mantlemint away from being just a faster LCD. In the current state, it already has the `indexing` feature that is not part of the standard LCD interface.
> 
> Sorry what do you mean by the `indexing` feature? I might not know about those endpoints

Hi @etienne-napoleone 
Mantlemint doesn't have tendermint and related API but it does index transactions and serve them.

Please refer to https://github.com/terra-money/mantlemint#default-indexes

File: export/export.go (hunk 5 above)
Comment (javiersuweijie) [reply export/export.go:78]: hmm. i think we should include them. let me add them.

**Review by jeffwoooo — APPROVED** [review]

**Comment by gregnuj** [comment]
Why are we attempting to merge this to phoenix-1 before it is deployed to testnet (pisco-1)

**Comment by jeffwoooo** [comment]
Let's merge into main branch for pull request, and fast-forward phoenix-1 or pisco-1 branch to deploy onto their node.

**Pull request merged**